### PR TITLE
Don't update config of components when tls is enable

### LINKED ISSUE #4298: `config` field is updated in TidbCluster
## Bug Report

**What version of Kubernetes are you using?**

**What version of TiDB Operator are you using?**

**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**

**What's the status of the TiDB cluster pods?**

**What did you do?**

Deploy a basic TiDB cluster with TLS enabled
**What did you expect to see?**
The `spec` part in TidbCluster CR is not updated
**What did you see instead?**
The config field is updated as below:
```
  pd:
    baseImage: pingcap/pd
    config: |
      [dashboard]
        internal-proxy = true
        tidb-cacert-path = "/var/lib/tidb-client-tls/ca.crt"
        tidb-cert-path = "/var/lib/tidb-client-tls/tls.crt"
        tidb-key-path = "/var/lib/tidb-client-tls/tls.key"

      [security]
        cacert-path = "/var/lib/pd-tls/ca.crt"
        cert-path = "/var/lib/pd-tls/tls.crt"
        key-path = "/var/lib/pd-tls/tls.key"
```
The certs fields are configured by TiDB Operator automatically, however, they should be set to the PD configurations file without updating the TidbCluster spec except for the default configurations we set explicitly.

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Close  #4298
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 

Please help test the following cases and make sure they are covered in e2e:

- Deploy TiDB cluster including PD, TiDB, TiKV, TiFlash, TiCDC, Pump
  - TLS not enabled
    - [x] Config for each component (TiFlash has two config fields) is `nil` (please note that some components like PD, TiDB, etc. cannot start with this config, just check the components that support this config)
    - [x] Config for each component (TiFlash has two config fields) is `{}`
    - [x] Config for each component (TiFlash has two config fields) has the valid config
  - TLS enabled
    - [x] Config for each component (TiFlash has two config fields) is `nil` (please note that some components like PD, TiDB, etc. cannot start with this config, just check the components that support this config)
    - [x] Config for each component (TiFlash has two config fields) is `{}`
    - [x] Config for each component (TiFlash has two config fields) has the valid config
- Deploy DM cluster including DM Master and DM worker (Same cases for TiDB cluster)
  - [x] TLS not enabled
  - [x] TLS enabled

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/manager/member/pump_member_manager.go [hunk 1]
```diff
@@ -284,23 +284,29 @@ func getNewPumpHeadlessService(tc *v1alpha1.TidbCluster) *corev1.Service {
 
 // getNewPumpConfigMap returns a configMap for pump
 func getNewPumpConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
-	if tc.Spec.Pump == nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/pump_member_manager.go:287]: Why remove this?

File: pkg/manager/member/tikv_member_manager.go [hunk 2]
```diff
@@ -702,8 +702,7 @@ func transformTiKVConfigMap(srcStr string, tc *v1alpha1.TidbCluster) string {
 }
 
 func getTikVConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
-	config := tc.Spec.TiKV.Config
-	if config == nil {
+	if tc.Spec.TiKV.Config == nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go:705]: Please also check `getTikVConfigMapForTiKVSpec`.

File: pkg/manager/member/pd_member_manager.go [hunk 3]
```diff
@@ -812,10 +811,10 @@ func getNewPDSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (
 
 func getPDConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 	// For backward compatibility, only sync tidb configmap when .tidb.config is non-nil
-	config := tc.Spec.PD.Config
-	if config == nil {
+	if tc.Spec.PD.Config == nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/pd_member_manager.go:814]: What about TiCDC and TiFlash?

**Comment by DanielZhangQD** [comment]
Please help test the following cases and make sure they are covered in e2e:

- Deploy TiDB cluster including PD, TiDB, TiKV, TiFlash, TiCDC, Pump
  - TLS not enabled
    - Config for each component (TiFlash has two config fields) is `nil` (please note that some components like PD, TiDB, etc. cannot start with this config, just check the components that support this config)
    - Config for each component (TiFlash has two config fields) is `{}`
    - Config for each component (TiFlash has two config fields) has the valid config
  - TLS enabled
    - Same cases with TLS not enabled
- Deploy DM cluster including DM Master and DM worker (Same cases for TiDB cluster)
  - TLS not enabled
  - TLS enabled

File: pkg/manager/member/pump_member_manager.go (hunk 1 above)
Comment (KanShiori) [reply pkg/manager/member/pump_member_manager.go:287]: I think the field should be verified before calling the function, just like other `getXXXConfigMap`

File: pkg/manager/member/pd_member_manager.go (hunk 3 above)
Comment (KanShiori) [reply pkg/manager/member/pd_member_manager.go:814]: seem to not need to config `[security]` when tls is enable

Comment (KanShiori) [reply pkg/manager/member/pd_member_manager.go:814]: ticdc use cmd args without config, and tiflash have been used `DeepCopy()`

**Comment by KanShiori** [comment]
> Please help test the following cases and make sure they are covered in e2e:
> 
> * Deploy TiDB cluster including PD, TiDB, TiKV, TiFlash, TiCDC, Pump
>   
>   * TLS not enabled
>     
>     * Config for each component (TiFlash has two config fields) is `nil` (please note that some components like PD, TiDB, etc. cannot start with this config, just check the components that support this config)
>     * Config for each component (TiFlash has two config fields) is `{}`
>     * Config for each component (TiFlash has two config fields) has the valid config
>   * TLS enabled
>     
>     * Same cases with TLS not enabled
> * Deploy DM cluster including DM Master and DM worker (Same cases for TiDB cluster)
>   
>   * TLS not enabled
>   * TLS enabled

I have tested these cases, all cases are successful.

File: pkg/manager/member/pd_member_manager.go (hunk 3 above)
Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager.go:814]: TiCDC also supports `config` from v1.2.x.

Comment (KanShiori) [reply pkg/manager/member/pd_member_manager.go:814]: Done

**Comment by DanielZhangQD** [comment]
@KanShiori Could you please update the TLS case https://github.com/pingcap/tidb-operator/blob/master/tests/e2e/tidbcluster/tidbcluster.go#L1112 here, to set configs for all components of TidbCluster and check that the configs are not updated?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**